### ISSUE TITLE
Search: Instant Search is still functional after subscription expiry date

### ISSUE THREAD

**Issue by kangzj** [Automattic/jetpack#20502]
#### Steps to reproduce the issue

1. Go to a site with __expired__ Jetpack Search Subscription - I don't have one like this, and just changed the expiry date from store admin.
2. Navigate to performance settings `/wp-admin/admin.php?page=jetpack#/performance`

#### What I expected
Instant Search should be shown as expired and provide a link for user to renew.

#### What happened instead
Search and Instant Search toggles are still available on the dashboard and so is the frontend (Instant Search is still working)

**Screenshots**
![image](https://user-images.githubusercontent.com/1425433/127596440-69d399a4-3970-4ad5-9dd2-26e1f3979446.png)

![image](https://user-images.githubusercontent.com/1425433/127596237-1a98788e-39e0-4bcd-8e8a-4f54032611f7.png)

If user doesn't have a Search subscription, the experience is all good tho.
![image](https://user-images.githubusercontent.com/1425433/127596660-6b20fad4-1eda-4cf7-938a-4f5418ada88c.png)

### Initial Investigation
When the frontend detected there's an active Jetpack Search subscription, it set an option which acts as cache and used by the backend to determine whether Jetpack supports search. The problem is I don't see anywhere to set this option to `false`. 
https://github.com/Automattic/jetpack/blob/d9f60dfa546f2302aa8a6329d2da71eaa5188a86/projects/plugins/jetpack/_inc/client/performance/search.jsx#L66
https://github.com/Automattic/jetpack/blob/d9f60dfa546f2302aa8a6329d2da71eaa5188a86/projects/plugins/jetpack/class.jetpack-plan.php#L319

And also the Elasticsearch API need to be checked as well to see why it still returns results even when subscription is expired - could be a cache somewhere.

#### Last but not least
Just a big guess here - this could potentially contribute to the subscription declining puzzle `pa1C6h-ZJ-p2` because if I can still use it why do I spend money renewing?

Also the fix could affect the search usability of a number of sites - probably we want to inform HE in advance before shipping the fix.

**Comment by kangzj** [comment]
### TL;DR
This is a false alarm. Phew...

Thanks for @jsnmoon's input for the issue. In the function on `wp-content/lib/jetpack-plans/search.php:161`, `has_jetpack_search_product` is set when Search subscription expires. 

And the search subscription is not available now on my test site anymore, so I believe there's a scheduled (or just async) job or sth which manages the deactivation of subscription, and calls Jetpack site endpoints in the code mentioned above.

However, if the Jetpack site is offline at the moment of the deactivation, Search might not be deactivated properly. Fortunately the performance dashboard calls wpcom to check whether there's a subscription and acts as expected. But `Jetpack_Plan::supports('search')` would still return true which would potentially cause issues.

### Conclusion
- There's some time gap after the expiry date is set and before Instant Search subscription goes unavailable.
- We'll need to ask the mechanism ( timing/retries ) of Production activation / deactivation and confirm if we need to worry about the problem when API called on offline JP sites.  

cc @bluefuton

**Comment by mdbitz** [comment]
I've been reviewing the subscription data and the follow flow is triggered if the site is not accessible by WP.com when a subscription for Search expires or is removed.

1) WP.com attempt to send deactivate API call fails
2) Jetpack performs a daily heartbeat which triggers a refresh of the Subscription data from WP.com
3) This update disables the search module but does not update the 'instant_search_enabled' or 'has_jetpack_search_product' option

To resolve this gap we need to add a hook so that after the plan or product options of the site update that we correctly update the options to false if they don't have a Complete plan or a Search Product.

I see a few ways of handling this:

1) Add a hook to the Jetpack_Plan [get](https://github.com/Automattic/jetpack/blob/374f9afcb435e9480463dba7959fa408a68c4672/projects/plugins/jetpack/class.jetpack-plan.php#L221) function. This loads the plan details and caches it, We could check for the state where Search isn't part of the plan/product and then force the options to be disabled. 

This would be extra check/read of option once per request.

2) Add a late hook to the heartbeat [action](https://github.com/Automattic/jetpack/blob/40b34101caa5d07a15527a737769ac87576fd89c/projects/plugins/jetpack/class.jetpack.php#L762) (higher than 10 priority). This would allow search to check the updated plan details and determine if the options should be disabled.

3) Add a hook looking at the update_option_jetpack_active_plan and update_option_jetpack_site_products. As we need to check both this would mean looking at the other option to check value and double processing. 

I think going ahead with #2 is appropriate. the hook should reside in Search Package and we'll also need to update some of the Plan logic so that it isn't circular based on these options. 
e.g. https://github.com/Automattic/jetpack/blob/f7e6d164278b58f37de4e8a6aab87c3187955532/projects/plugins/jetpack/modules/search/class.jetpack-search.php#L235 is okay as long as we have a way to check the plan /products directly to validate the [option](https://github.com/Automattic/jetpack/blob/374f9afcb435e9480463dba7959fa408a68c4672/projects/plugins/jetpack/class.jetpack-plan.php#L322) is accurate.

**Comment by kangzj** [comment]
Hi @mdbitz thank you so much for your input 👍 . I'm currently working on this. Because search is going standalone, we are implementing its own methods for plan supports determination and etc, including:

- An endpoint for all search subscription ( D69004 )
- A consumer in Jetpack to cache and provide calls based on the data (#21584)
    - I'm looking for the right timing to update the plan data, the heartbeat action is definitely a great choice. Thanks heaps for the suggestions - that's definitely so much better than what it is now.
    - I'm planning to remove `has_jetpack_search_product` in the future (it will stick around for compatibility for a while tho), and change it to use `Automattic\Jetpack\Search\Plan` ([code](https://github.com/Automattic/jetpack/blob/master/projects/packages/search/src/class-plan.php)).
- I'm migrating the process of activating and deactivating Search on JP sites as well and will keep everyone posted on how it goes.

Thanks again.

**Comment by mdbitz** [comment]
Excellent, Thanks for the update :)

**Comment by kangzj** [comment]
The issue should be gone as we implemented a new [Plan](https://github.com/Automattic/jetpack/blob/2ef743a8e4a3f4012a24b663020a0b0da2c01dfd/projects/packages/search/src/class-plan.php#L44) class which update the eligibility info on Search Board loading and Jetpack heartbeat. More info could be found here: `pcNPJE-FG-p`.
